Force inline Timer::attach() to get rid of floating-point instructions

hugueskamba opened the pull request:
### Description
Calls to Timer::attach() are inlined in order not to use floating-point
library functions calls given Timer::attach_us() expects an integer
for the callback interval.

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@evedon @gpsimenos @jamesbeyond 
<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

hugueskamba committed:
Require integer to specify callback interval with Ticker API

* Create new `s_timestamp_t` type to specify timestamp is seconds
* Alter `attach()` API to expect `s_timestamp_t` for interval value
* Create helper macro to convert seconds to milliseconds and help code
  readability
* Modify Greentea tests accordingly

evedon commented on drivers/Ticker.h:
```diff
@@ -78,9 +78,9 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
      *  @param t the time between calls in seconds
      */
-    void attach(Callback<void()> func, float t)
+    void attach(Callback<void()> func, const s_timestamp_t t)
```
You should deprecate the current API and provide a new one which takes s_timestamp_t

kjbracey replied:
Indeed, this changes the behaviour of existing code. `attach(0.5)` will be treated as `attach(0)`. 

You should get a warning, but that could be missed.

hugueskamba replied:
c86ac7e

kjbracey replied:
Unfortunately, this override won't really work, unless I'm mistaken.

I believe `attach(2)` will be ambiguous - `2` is neither a `uint64_t` nor a `float`, and it won't know which overload to prefer.

We could add overrides for `std::duration`, as I suggested below, but not for more implicitly-convertable arithmetic types :(

To deprecate, it would have to be an actual new API like `attach_s`.

hugueskamba replied:
The change I made make the user more conscious of what s/he is doing. S/he will have to deliberately either cast `2` to `s_timestamp_t` or define a variable of type `s_timestamp_t`. This is a good thing. Passing just `2` causes a compilation error.

kjbracey replied:
Granted, that's better than a silent behaviour change, but making existing code not compile is a lot stronger than a deprecation, so I'm not sure whether it's merited.

(But it's also not strongly typed either - they could pass a `us_timestamp_t`, and it wouldn't bat an eyelid.)

evedon replied:
attach_s is a good suggestion since we already have attach_us.

hugueskamba replied:
Ok I went for `Ticker::attach_s()`, 1709847,  be933a6

evedon replied:
That's great.

evedon commented on hal/ticker_api.h:
```diff
@@ -24,6 +24,21 @@
+/**
+ * Converts microseconds to seconds
+ */
+#define MICROSECONDS_TO_SECONDS(MICROSECONDS) (s_timestamp_t)(MICROSECONDS / MICROSECONDS_IN_SECOND)
```
Define the macros where they are needed. 
In addition, only SECONDS_TO_MICROSECONDS is used in actual code (not considering test code).

hugueskamba replied:
It was defined where typedef related to timing are defined to be accessible along with all timing stuff.
Even if it is only used in test code SECONDS_TO_MICROSECONDS is a nice abstraction that applications can also use. The conversion from seconds to microseconds happened two places so using the macro makes sense. It will avoid users in the future doing a conversion and using `1000000f` again.

hugueskamba replied:
If you feel strongly about the location of the macros, 49e1c52

evedon replied:
Great, thanks.

hugueskamba committed:
Deprecate previous Ticker::attach() API

* Restore previous implementation as deprecated
* Ensure method call is unambiguous

hugueskamba committed:
Move time macro conversions to `Ticker.h`

hugueskamba committed:
Add new API method `attach_s` instead of overloading the deprecated `attach`

hugueskamba committed:
Ensure tests use the new method.

evedon commented on TESTS/mbed_drivers/lp_ticker/main.cpp:
```diff
@@ -142,7 +142,7 @@ void test_detach(void)
     LowPowerTicker ticker;
     bool ret;
-    const float ticker_time_s = 0.1f;
+    const s_timestamp_t ticker_time_s = 1;
```
If this is now one second instead of 0.1s then you should increase wait_time_ms

hugueskamba replied:
The wait time has been incremented by the same factor. d435b1d

evedon commented on TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -264,7 +264,7 @@ void test_detach(void)
     Ticker ticker;
     bool ret;
-    const float ticker_time_s = 0.1f;
+    const s_timestamp_t ticker_time_s = 1;
```
Same comment here

hugueskamba replied:
d435b1d

hugueskamba committed:
Fix lp_ticker and ticker unit test case wait times

hugueskamba committed:
Fix code indentation to satisfy code style checker

hugueskamba committed:
Ensure the new method is called by the deprecated one

hugueskamba committed:
Force inline Timer::attach() to get rid of floating-point instruction

Calls to Timer::attach() are inlined in order not to use floating-point
library functions calls given Timer::attach_us() expects an integer
for the callback interval.

evedon reviewed (approved):
Nice solution indeed !

hugueskamba committed:
Use Perfect forwading to pass arguments from `attach()` to `attach_us()`

This prevents a possible heavy callback copy.

kjbracey commented on drivers/Ticker.h:
```diff
@@ -75,10 +75,15 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
      */
-    void attach(Callback<void()> func, float t)
+    template <typename F, typename T>
+    MBED_FORCEINLINE void attach(F&& func, T&& t)
```
Hmm, hadn't occurred to me to templatise the actual time. I can't immediately see a particular benefit to that. Actually, it will get in the way of future `std::duration` overloads, so it's not acceptable.

If you really wanted it, you'd have to qualify it with an `enable_if<is_arithmetic>` or something, but I think templatising that parameter is just complication for no reason. (You're not just forwarding the time parameter - you're working on it).

Any reason to omit the `std::forward<F>`? Even if it doesn't make a difference now, it should make a difference with some upcoming `Callback` optimisations.

hugueskamba replied:
I un-templatized the time interval. 
573d8c6
See the size difference below.
Thanks

With reference forwarding

```
| Module             |      .text |    .data |      .bss |
|--------------------|------------|----------|-----------|
| [fill]             |    157(+0) |    2(+0) |  2187(+0) |
| [lib]/c.a          |   6892(+0) | 2108(+0) |    89(+0) |
| [lib]/gcc.a        |    772(+0) |    0(+0) |     0(+0) |
| [lib]/misc         |    180(+0) |    4(+0) |    28(+0) |
| main.o             |  13686(+0) |    0(+0) |  1337(+0) |
| mbed-os/components |     34(+0) |    0(+0) |     0(+0) |
| mbed-os/drivers    |    856(+0) |    0(+0) |     0(+0) |
| mbed-os/features   | 109465(+0) |    8(+0) |    12(+0) |
| mbed-os/hal        |   1560(+0) |    4(+0) |    67(+0) |
| mbed-os/platform   |   4877(+0) |  262(+0) |   478(+0) |
| mbed-os/rtos       |   6570(+0) |  168(+0) |  5973(+0) |
| mbed-os/targets    |   5586(+0) |   12(+0) |   381(+0) |
| Subtotals          | 150635(+0) | 2568(+0) | 10552(+0) |
Total Static RAM memory (data + bss): 13120(+0) bytes
Total Flash memory (text + data): 153203(+0) bytes
```


Without reference forwarding
```
| Module             |        .text |    .data |      .bss |
|--------------------|--------------|----------|-----------|
| [fill]             |      161(+4) |    2(+0) |  2187(+0) |
| [lib]/c.a          |     6892(+0) | 2108(+0) |    89(+0) |
| [lib]/gcc.a        |      772(+0) |    0(+0) |     0(+0) |
| [lib]/misc         |      180(+0) |    4(+0) |    28(+0) |
| main.o             |  14100(+414) |    0(+0) |  1337(+0) |
| mbed-os/components |       0(-34) |    0(+0) |     0(+0) |
| mbed-os/drivers    |      856(+0) |    0(+0) |     0(+0) |
| mbed-os/features   |   109465(+0) |    8(+0) |    12(+0) |
| mbed-os/hal        |     1560(+0) |    4(+0) |    67(+0) |
| mbed-os/platform   |     4877(+0) |  262(+0) |   478(+0) |
| mbed-os/rtos       |     6570(+0) |  168(+0) |  5973(+0) |
| mbed-os/targets    |     5586(+0) |   12(+0) |   381(+0) |
| Subtotals          | 151019(+384) | 2568(+0) | 10552(+0) |
Total Static RAM memory (data + bss): 13120(+0) bytes
Total Flash memory (text + data): 153587(+384) bytes
```

hugueskamba committed:
Un-templatise interval and forward the reference

Also fix code style to please style checker

kjbracey commented on drivers/Ticker.h:
```diff
@@ -75,12 +75,17 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
+    MBED_FORCEINLINE void attach(F &&func, float t)
     {
-        attach_us(func, t * 1000000.0f);
+        attach_us(std::forward<F>(func), t * 1000000.0f);
```
You should put the include for `forward` in.

ARMC5 compatibility needs it to be `#include <mstd_utility>`, rather than just `#include <utility>`

(See https://github.com/ARMmbed/mbed-os/blob/master/platform/cxxsupport/README.md - I need to get on with the handbook page for that).

hugueskamba replied:
Some points first:
1. Without using what you are suggesting above, the code compiles fine for ARMC5.
2. With or without what you are suggesting there is still one floating point instruction (only for ARMC5, have not looked at where it comes from)

Now the questions:
1. Is it worth doing given we are moving to ARMC6?
2. Are you suggesting adding 2 conditional pre-processor directives in Ticker.h as follows?
```
#if defined(__CC_ARM)
#include <mstd_utility>
#endif
...

    template <typename F>
    MBED_FORCEINLINE void attach(F &&func, float t)
    {
        attach_us(
#if defined(__CC_ARM)
            mstd::forward<F>(func),
#else
            std::forward<F>(func),
#endif // defined(__CC_ARM)
            t * 1000000.0f
        );
    }
```

kjbracey replied:
1. We're trying not to break the ARMC5 build outright for now, which is what that cxxsupport directory is partly about. It should be minimal effort from most people - I'm taking the hit. There's no need for `#ifdef`s anywhere in code - all the ifdeffing happens in the cxxsupport headers, but you just need to make sure to use `<mstd_utility>`, to give us the chance to get in front of the ARMC5's built-in `<utility>`

1. No, the only change is to add `#include <mstd_utility>`, unconditionally. That works for all toolchains. You can keep it as `std::forward`.

hugueskamba replied:
d4497b7

hugueskamba committed:
Include `mstd_utility` to get in front of the ARMC5's built-in `utility`

hugueskamba committed:
Rename hal/rtc test callback to avoid overloading
